src/contracts/exchange/NiftyswapExchange20.sol

```diff
@@ -0,0 +1,812 @@
+
+      // Current total liquidity calculated in currency token
+      uint256 totalLiquidity = totalSupplies[tokenId];
+
+      // When reserve for this token already exists
+      if (totalLiquidity > 0) {
+
+        // Load currency token and Token reserve's supply of Token id
+        uint256 currencyReserve = currencyReserves[tokenId]; // Amount not yet in reserve
+        uint256 tokenReserve = tokenReserves[i];
+
+        /**
+        * Amount of currency tokens to send to token id reserve:
+        * X/Y = dx/dy
+        * dx = X*dy/Y
+        * where
+        *   X:  currency total liquidity
+        *   Y:  Token _id total liquidity (before tokens were received)
+        *   dy: Amount of token _id deposited
+        *   dx: Amount of currency to deposit
+        *
+        * Adding .add(1) if rounding errors so to not favor users incorrectly
+        */
+        (uint256 currencyAmount, bool rounded) = divRound(amount.mul(currencyReserve), tokenReserve.sub(amount));
+        require(_maxCurrency[i] >= currencyAmount, "NiftyswapExchange#_addLiquidity: MAX_CURRENCY_AMOUNT_EXCEEDED");
+
+        // Update currency reserve size for Token id before transfer
+        currencyReserves[tokenId] = currencyReserve.add(currencyAmount);
+
+        // Update totalCurrency
+        totalCurrency = totalCurrency.add(currencyAmount);
+
+        // Proportion of the liquidity pool to give to current liquidity provider
+        // If rounding error occured, round down to favor previous liquidity providers
+        // See https://github.com/0xsequence/niftyswap/issues/19
+        liquiditiesToMint[i] = (currencyAmount.sub(rounded ? 1 : 0)).mul(totalLiquidity) / currencyReserve;
+        currencyAmounts[i] = currencyAmount;
+
+        // Mint liquidity ownership tokens and increase liquidity supply accordingly
+        totalSupplies[tokenId] = totalLiquidity.add(liquiditiesToMint[i]);
+
+      } else {
+        uint256 maxCurrency = _maxCurrency[i];
+
+        // Otherwise rounding error could end up being significant on second deposit
+        require(maxCurrency >= 1000000000, "NiftyswapExchange#_addLiquidity: INVALID_CURRENCY_AMOUNT");
+
+        // Update currency  reserve size for Token id before transfer
+        currencyReserves[tokenId] = maxCurrency;
+
+        // Update totalCurrency
+        totalCurrency = totalCurrency.add(maxCurrency);
+
+        // Initial liquidity is amount deposited (Incorrect pricing will be arbitraged)
+        // uint256 initialLiquidity = _maxCurrency;
+        totalSupplies[tokenId] = maxCurrency;
+
+        // Liquidity to mints
+        liquiditiesToMint[i] = maxCurrency;
+        currencyAmounts[i] = maxCurrency;
+      }
+    }
+
+    // Mint liquidity pool tokens
+    _batchMint(_provider, _tokenIds, liquiditiesToMint, "");
+
+    // Transfer all currency to this contract
+    currency.transferFrom(_provider, address(this), totalCurrency);
```

Comment:
Should require that this returns true since not all ERC-20 transfer implementation will throw on error.